index/repo_fs.go

```diff
@@ -54,12 +52,30 @@ func NewLocalIndexRepo(factory FullIndexFactory, baseDir string) (*LocalIndexRep
 	return l, nil
 }
 
-func (l *LocalIndexRepo) GetFullIndex(key shard.Key) (idx index.FullIndex, err error) {
-	// Use the factory to build a FullIndex from the file at the key path
-	return l.factory.Build(l.indexPath(key))
+func (l *FSIndexRepo) GetFullIndex(key shard.Key) (FullIndex, error) {
+	idx := NewMockFullIndex() // TODO replace with an implementation backed by Carbs serialization/deserialization.
+	path := l.indexPath(key)
+
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		closeErr := f.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	err = idx.Unmarshal(f)
+	if err != nil {
+		return nil, err
+	}
+	return idx, nil
```

Comment:
I implemented it the way you have it here originally, but the problem is that this assumes that the entire index is in memory.
If instead we use a factory pattern, we can just give the index file path to the index implementation, and the implementation can read / seek over the file as needed without needing to load the entire index into memory. This also gives full control to the index over when to open and close the file handle.